app/controllers/uploads_controller.rb

```diff
@@ -270,6 +301,186 @@ def complete_external_upload
     end
   end
 
+  def create_multipart
+    return render_404 if !SiteSetting.enable_direct_s3_uploads
+
+    RateLimiter.new(
+      current_user, "create-multipart-upload", CREATE_MULTIPART_RATE_LIMIT_PER_MINUTE, 1.minute
+    ).performed!
+
+    file_name = params.require(:file_name)
+    file_size = params.require(:file_size).to_i
+    upload_type = params.require(:upload_type)
+    content_type = params.require(:content_type)
```

Comment:
Do we need to validate the user input here as well? Will anything bad happen if someone passes in random bad strings?

Reply:
I don't think anything bad will happen, but I changed this to guess the content type using MiniMime which should be reliable:

https://github.com/discourse/discourse/blob/25c8817e61acad3634e4a6e31088b62a355efcf4/app/controllers/uploads_controller.rb#L563-L563

If the content type is blank nothing bad happens either; it's just a hint to S3 about what the content-type of the object should be